crates/rwasm/machine/src/operations/range.rs

```diff
@@ -145,4 +145,176 @@ impl<T: Copy, const START: u32, const END: u32> Range16bCols<T, START, END> {
+        builder.assert_eq(do_check, is_real);
+        Range16bCols::<AB::Var, START, END>::range_check(builder, cols);
+    }
+}
+
+#[derive(AlignedBorrow, Default, Debug, Clone, Copy)]
+#[repr(C)]
+pub struct Range32bCols<
+    T,
+    const START_HI16: u32,
+    const START_LOW16: u32,
+    const END_HI16: u32,
+    const END_LOW16: u32,
+> {
+    hi_16bits: Range16bCols<T, START_HI16, END_HI16>,
+    low_16bits: Range16bCols<T, START_LOW16, END_LOW16>,
+    pub hi_is_zero: T,
+    pub hi_is_not_edge: T,
+    pub hi_is_eq_ub_hi: T,
+}
+
+impl<
+        F: PrimeField32,
+        const START_HI16: u32,
+        const START_LOW16: u32,
+        const END_HI16: u32,
+        const END_LOW16: u32,
+    > Range32bCols<F, START_HI16, START_LOW16, END_HI16, END_LOW16>
+{
+    pub fn populate_value(&mut self, value: u32) {
+        // We subtract the START to work with the value that is in the range [0..END -
+        // START]
+        let start = START_HI16 << 16 + START_LOW16;
+
+        let (shifted_value, overflow) = value.overflowing_sub(START_HI16);
+
+        assert!(!overflow);
+
+        let shifted_value: u32 = shifted_value.try_into().unwrap();
+
+        let hi_16bits: u16 = (shifted_value >> 16) as u16;
+        let low_16bits: u16 = shifted_value as u16;
+        self.hi_16bits.populate_value(hi_16bits as u32);
+        self.low_16bits.populate_value(low_16bits as u32);
+    }
+
+    pub fn populate(&mut self, value: u32, output: &mut impl ByteRecord) {
+        // We subtract the START to work with the value that is in the range [0..END -
+        // START]
+        let start = START_HI16 << 16 + START_LOW16;
+        let (shifted_value, overflow) = value.overflowing_sub(start);
+
+        assert!(!overflow);
+        let hi_16bits: u16 = (shifted_value >> 16) as u16;
+        let low_16bits: u16 = shifted_value as u16;
+
+        let hi_is_eq_ub_hi = hi_16bits == Self::UB_HI_16BITS_SHIFTED;
+        self.hi_is_eq_ub_hi = F::from_bool(hi_is_eq_ub_hi);
+
+        let hi_is_zero = hi_16bits == 0;
+        self.hi_is_zero = F::from_bool(hi_is_zero);
+
+        let hi_is_not_edge = !hi_is_zero && !hi_is_eq_ub_hi;
+        self.hi_is_not_edge = F::from_bool(hi_is_not_edge);
+
+        output.add_u16_range_check(hi_16bits);
+        output.add_u16_range_check(low_16bits);
+        if hi_is_eq_ub_hi {
+            self.low_16bits.populate(low_16bits as u32, output);
+        } else {
+            self.low_16bits.populate_value(low_16bits as u32);
+        }
+        if hi_is_not_edge {
+            self.hi_16bits.populate(hi_16bits as u32, output);
+        } else {
+            self.hi_16bits.populate_value(hi_16bits as u32);
+        }
+    }
+}
+
+impl<
+        T: Copy,
+        const START_HI16: u32,
+        const START_LOW16: u32,
+        const END_HI16: u32,
+        const END_LOW16: u32,
+    > Range32bCols<T, START_HI16, START_LOW16, END_HI16, END_LOW16>
+{
+    const START: u32 = (START_HI16 << 16 + START_LOW16);
```

Comment:
It's strange to calculate the value from generic high/low rather than vice versa